inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/util/MetricStateUtils.java

```diff
@@ -0,0 +1,127 @@
+import org.apache.flink.api.common.state.ListState;
+import org.apache.inlong.sort.base.metric.MetricState;
+import org.apache.inlong.sort.base.metric.SourceMetricData;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.apache.inlong.sort.base.Constants.NUM_BYTES_IN;
+import static org.apache.inlong.sort.base.Constants.NUM_RECORDS_IN;
+
+/**
+ * metric state for {@link MetricState} supporting snapshot and restore
+ */
+@Slf4j
+public class MetricStateUtils {
+
+    /**
+     *
+     * restore metric state data
+     * @param metricStateListState state data list
+     * @param subtaskIndex current subtask index
+     * @param currentSubtaskNum number of current parallel subtask
+     * @return metric state
+     * @throws Exception throw exception metricStateListState.get()
+     */
+    public static MetricState restoreMetricState(ListState<MetricState> metricStateListState, Integer subtaskIndex,
+            Integer currentSubtaskNum) throws Exception {
+        if (metricStateListState == null || metricStateListState.get() == null) {
+            return null;
+        }
+        log.info("restoreMetricState:{}, subtaskIndex:{}, currentSubtaskNum:{}", metricStateListState, subtaskIndex,
+                currentSubtaskNum);
+        MetricState currentMetricState;
+        Map<Integer, MetricState> map = new HashMap<>(16);
+        for (MetricState metricState : metricStateListState.get()) {
+            map.put(metricState.getSubtaskIndex(), metricState);
+        }
+        int previousSubtaskNum = map.size();
+        if (currentSubtaskNum >= previousSubtaskNum) {
+            currentMetricState = map.get(subtaskIndex);
+        } else {
+            Map<String, Long> metrics = new HashMap<>(4);
+            currentMetricState = new MetricState(subtaskIndex, metrics);
+            List<Integer> indexList = computeIndexList(subtaskIndex, currentSubtaskNum, previousSubtaskNum);
+            for (Integer index : indexList) {
+                MetricState metricState = map.get(index);
+                for (Map.Entry<String, Long> entry : metricState.getMetrics().entrySet()) {
+                    if (metrics.containsKey(entry.getKey())) {
+                        metrics.put(entry.getKey(), metrics.get(entry.getKey()) + entry.getValue());
+                    } else {
+                        metrics.put(entry.getKey(), entry.getValue());
+                    }
+                }
+            }
+        }
+        return currentMetricState;
+    }
+
+    /**
+     *
+     * Assignment previous subtask index to current subtask when reduce parallelism
+     * @param subtaskIndex current subtask index
+     * @param currentSubtaskNum number of current parallel subtask
+     * @param previousSubtaskNum number of previous parallel subtask
+     * @return index list
+     */
+    public static List<Integer> computeIndexList(Integer subtaskIndex, Integer currentSubtaskNum,
```

Comment:
the doc should explain the rule on how are the indexs are assigned

Reply:
add description